lib/go-tc/servers.go

```diff
@@ -49,6 +64,157 @@ type ServersV1DetailResponse struct {
+// ServerInterfaceInfo structures. No interfaces will be marked for monitoring,
+// and it will generate service addresses according to the passed indicators
+// for each address family.
+func (lid *LegacyInterfaceDetails) ToInterfaces(ipv4IsService, ipv6IsService bool) ([]ServerInterfaceInfo, error) {
+	var iface ServerInterfaceInfo
+	if lid.InterfaceMtu == nil {
+		return nil, errors.New("interfaceMtu is null")
+	}
+	mtu := uint64(*lid.InterfaceMtu)
+	iface.MTU = &mtu
+
+	if lid.InterfaceName == nil {
+		return nil, errors.New("interfaceName is null")
+	}
+	iface.Name = *lid.InterfaceName
+
+	var ips []ServerIpAddress
+	if lid.IPAddress != nil && *lid.IPAddress != "" {
+		if lid.IPGateway != nil && *lid.IPGateway == "" {
+			lid.IPGateway = nil
+		}
+
+		ipStr := *lid.IPAddress
+		if lid.IPNetmask != nil && *lid.IPNetmask != "" {
+			mask := net.ParseIP(*lid.IPNetmask).To4()
+			if mask == nil {
+				return nil, fmt.Errorf("Failed to parse netmask '%s'", *lid.IPNetmask)
+			}
+			cidr, _ := net.IPv4Mask(mask[0], mask[1], mask[2], mask[3]).Size()
+			ipStr = fmt.Sprintf("%s/%d", ipStr, cidr)
+		}
+
+		ips = append(ips, ServerIpAddress{
+			Address:        ipStr,
+			Gateway:        lid.IPGateway,
+			ServiceAddress: ipv4IsService,
+		})
+	}
+
+	if lid.IP6Address != nil && *lid.IP6Address != "" {
+		if lid.IP6Gateway != nil && *lid.IP6Gateway == "" {
+			lid.IP6Gateway = nil
+		}
+		ips = append(ips, ServerIpAddress{
+			Address:        *lid.IP6Address,
+			Gateway:        lid.IP6Gateway,
+			ServiceAddress: ipv6IsService,
+		})
+	}
+
+	iface.IPAddresses = ips
+	return []ServerInterfaceInfo{iface}, nil
+}
+
+// InterfaceInfoToLegacyInterfaces converts a ServerInterfaceInfo to an
+// equivalent LegacyInterfaceDetails structure. It does this by creating the
+// IP address fields using the "service" interface's IP addresses. All others
+// are discarded, as the legacy format is incapable of representing them.
+func InterfaceInfoToLegacyInterfaces(serverInterfaces []ServerInterfaceInfo) (LegacyInterfaceDetails, error) {
+	var legacyDetails LegacyInterfaceDetails
+
+	for _, intFace := range serverInterfaces {
+
+		for _, addr := range intFace.IPAddresses {
+			if !addr.ServiceAddress {
+				continue
+			}
+
+			address := addr.Address
+			gateway := addr.Gateway
+
+			var parsedIp net.IP
+			var mask *net.IPNet
+			var err error
+			parsedIp, mask, err = net.ParseCIDR(address)
+			if err != nil {
+				parsedIp = net.ParseIP(address)
+				if parsedIp == nil {
+					return legacyDetails, fmt.Errorf("Failed to parse '%s' as network or CIDR string: %v", address, err)
+				}
+			}
+
+			if parsedIp.To4() == nil {
+				legacyDetails.IP6Address = &address
+				legacyDetails.IP6Gateway = gateway
+			} else if mask != nil {
+				legacyDetails.IPAddress = util.StrPtr(parsedIp.String())
+				legacyDetails.IPGateway = gateway
+				legacyDetails.IPNetmask = util.StrPtr(fmt.Sprintf("%d.%d.%d.%d", mask.Mask[0], mask.Mask[1], mask.Mask[2], mask.Mask[3]))
```

Comment:
Wouldn't it be possible for the legacy interface to have the netmask field set for a different address/gateway?
E.g.
```
cache
  eth0 
     addr: x.y.w.z
     gateway: x.y.w.z
     netmask: x.y.w.z
  lo:
    addr: i,j,k,l
    gateway: i,j,k,l
```
resulting in a legacy interface that has the `lo` addr/gateway with the `eth0` netmask.

It seems to me that we should unset the IPNetmask in the else below.

Reply:
This shouldn't be an issue for well-formed `ServerInterfaceInfo`s because they can only legally contain one IPv4 service address, but that's probably still a good idea.